File: plot/plot.py
# plot.py
# Reads JSON data from a file or stdin and plots curves, circles, and points using Matplotlib.

import sys
import json
import matplotlib.pyplot as plt

def plot_from_json(data):
    for item in data:
        item_type = item.get('type')

        if item_type == "curve":
            x = item.get('x', [])
            y = item.get('y', [])
            plt.plot(x, y, label='Curve')

        elif item_type == "circle":
            center = item.get('center')
            radius = item.get('radius')
            circle = plt.Circle((center['x'], center['y']), radius, color='r', fill=False)
            plt.gca().add_patch(circle)
            plt.scatter(center['x'], center['y'], color='r', label='Circle Center')

        elif item_type == "point":
            x = item.get('x', [])
            y = item.get('y', [])
            plt.scatter(x, y, label='Points')

    plt.xlabel('X')
    plt.ylabel('Y')
    plt.title('JSON Data Plot')
    plt.legend()
    plt.show()

if __name__ == '__main__':
    if len(sys.argv) == 2:
        # If a command-line argument is provided, assume it's a JSON filename
        json_filename = sys.argv[1]
        with open(json_filename, 'r') as json_file:
            data = json.load(json_file)
            plot_from_json(data)
    else:
        # If no argument is provided, read JSON data from stdin
        json_data = sys.stdin.read()
        data = json.loads(json_data)
        plot_from_json(data)

File: smallest-circle/plot.lua
-- plot.lua
-- This module provides functions to add plot objects (points, curves, circles)
-- and generate plots using a Python script. It allows data to be saved to a
-- JSON file or passed directly through a pipe to the Python script.

local cjson = require("cjson")
local Plot = {}

-- Global variables for script and JSON filenames
Plot.SCRIPT_NAME = "plot.py"
Plot.JSON_NAME = "plot.json"

Plot.plot_data = {}

function Plot.addPoint(x, y, label)
  table.insert(Plot.plot_data, {
    type = "point",
    x = x,
    y = y,
    label = label
  })
end

function Plot.addCurve(x, y, label)
  table.insert(Plot.plot_data, {
    type = "curve",
    x = x,
    y = y,
    label = label
  })
end

function Plot.addCircle(center_x, center_y, radius, label, color)
  table.insert(Plot.plot_data, {
    type = "circle",
    center = {x = center_x, y = center_y},
    radius = radius,
    label = label,
    color = color,
  })
end

-- Save plot data to a file with default or specified name
function Plot.saveToFile(filename)
  filename = filename or Plot.JSON_NAME
  local json_data = cjson.encode(Plot.plot_data)
  local file = io.open(filename, "w")
  file:write(json_data)
  file:close()
end

function Plot.plot(use_file)
  if use_file then
    -- write to a file and call Python script with file
    Plot.saveToFile()
    local command = 'python "' .. Plot.SCRIPT_NAME .. '" ' .. Plot.JSON_NAME
    os.execute(command)
  else
    -- call Python script through a pipe with JSON data
    local json_data = cjson.encode(Plot.plot_data)
    local pipe = io.popen('python "' .. Plot.SCRIPT_NAME .. '"', 'w')
    pipe:write(json_data)
    pipe:close()
  end
end

return Plot
